Doc/library/stdtypes.rst

```diff
@@ -4822,35 +4822,67 @@ Generic Alias Type
+``food = ['spam', 'eggs', 'bacon']``, calling ``food[0]`` will return the same
+value as calling ``type(food).__getitem__(food, 0)``. However, if a class
+defines the classmethod :meth:`__class_getitem__<object.__class_getitem__>`,
+then the subscription of that class may instead call the class's implementation
+of ``__class_getitem__``. This should return a ``GenericAlias`` object if it is
+properly defined. For example, because the :class:`list` class defines
+``__class_getitem__``, calling ``list[str]`` is equivalent to calling
+``list.__class_getitem__(str)``, rather than
+``type(list).__getitem__(list, str)``.
 
 .. note::
-   If the :meth:`__getitem__` of the class' metaclass is present, it will take
-   precedence over the :meth:`__class_getitem__` defined in the class (see
-   :pep:`560` for more details).
-
-The ``GenericAlias`` object acts as a proxy for :term:`generic types
-<generic type>`, implementing *parameterized generics* - a specific instance
-of a generic which provides the types for container elements.
-
-The user-exposed type for the ``GenericAlias`` object can be accessed from
-:class:`types.GenericAlias` and used for :func:`isinstance` checks.  It can
-also be used to create ``GenericAlias`` objects directly.
+   If :meth:`__getitem__<object.__getitem__>` is defined by a class's
+   :term:`metaclass`, it will take precedence over a
+   :meth:`__class_getitem__<object.__class_getitem__>` classmethod defined by
+   the class. See :pep:`560` for more details.
+
+A ``GenericAlias`` object acts as a proxy for a :term:`generic type`,
+implementing *parameterized generics*.
+
+For a container class which implements
+:meth:`__class_getitem__<object.__class_getitem__>`, the argument(s) supplied
+to a :ref:`subscription<subscriptions>` of the class may indicate the type(s)
+of the elements an object contains. For example, ``list[int]`` can be used in
+type annotations to signify a :class:`list` in which all the elements are of
+type :class:`int`.
+
+For a class which defines :meth:`__class_getitem__<object.__class_getitem__>`
+but is not a container, the
+argument(s) supplied to a subscription of the class will often indicate the
+return type(s) of one or more methods defined on an object. For example,
+:mod:`regular expressions<re>` can be used on the :class:`str` data type and
+the :class:`bytes` data type. If ``x = re.search('foo', 'foo')``, ``x`` will be
+a :ref:`re.Match <match-objects>` object where the return values of
+``x.group(0)`` and ``x[0]`` will both be of type :class:`str`. We can represent
+this kind of object in type annotations with the ``GenericAlias``
+``re.Match[str]``. If ``y = re.search(b'bar', b'bar')``, however, ``y`` will
+also be an instance of ``re.Match``, but the return values of ``y.group(0)``
+and ``y[0]`` will both be of type :class:`bytes`. In type annotations, we would
+represent this subtype of :ref:`re.Match <match-objects>` with
+``re.Match[bytes]``.
```

Comment:
Can anybody think of a better example here? I'm not a massive fan of the example I've used here, because it requires some degree of knowledge of regular expressions, which is a corner of Python not everybody knows about. (This paragraph is also longer than I'd like it to be...)

Ideally, I'd like an example that satisfies the following criteria:

1. Is a class that is relatively easy to understand. That rules out ``os.PathLike`` (it's an ABC where the significance of ``__class_getitem__`` is the return value of ``__fspath__``) and ``functools.cached_property`` (don't want to have to explain descriptors!).
2. Is a class that is *obviously* "not a container". That rules out ``queue.Queue``, which feels "kinda like a container even though it technically isn't", and anything from ``collections.abc``. (``collections.abc.Iterable`` is not a container, for example, but I don't want to use an example from ``collections.abc``, as the title of the documentation for ``collections.abc`` is literally "[collections.abc — Abstract Base Classes for Containers](https://docs.python.org/3/library/collections.abc.html)".)

If anybody has any ideas of how to improve this bit, I'd love to hear them!

Reply:
Perhaps `difflib.SequenceMatcher`? The generic specifies the type of item in the sequences it compares. Even if you've not used the module, the name makes fairly obvious roughly what it does.

Otherwise a not as suitable alternative is `concurrent.futures.Future`. The generic specifies the result type contained within, describing it as holding that would possibly suffice. Just not go into detail about what it's used for.

Reply:
TIL there's a `difflib` module! I think I'll stick with the `re` example for now — I'm happier with it now that the example's been split into bullet points, it feels more manageable — but thanks 🙂